pageserver/src/remote_storage/storage_sync.rs

```diff
@@ -134,151 +143,123 @@ lazy_static! {
+    sender: UnboundedSender<(ZTenantTimelineId, SyncTask)>,
+    receiver: Mutex<UnboundedReceiver<(ZTenantTimelineId, SyncTask)>>,
+}
 
-    /// Adds a new task to the queue, if the queue was initialized, returning `true` on success.
-    /// On any error, or if the queue was not initialized, the task gets dropped (not scheduled) and `false` is returned.
-    pub fn push(sync_id: ZTenantTimelineId, new_task: SyncTask) -> bool {
-        if let Some(sender) = SENDER.get() {
-            match sender.send((sync_id, new_task)) {
-                Err(e) => {
-                    warn!("Failed to enqueue a sync task: the receiver is dropped: {e}");
-                    false
-                }
-                Ok(()) => {
-                    LENGTH.fetch_add(1, Ordering::Relaxed);
-                    true
-                }
-            }
-        } else {
-            warn!("Failed to enqueue a sync task: the sender is not initialized");
-            false
+impl SyncQueue {
+    fn new(max_timelines_per_batch: NonZeroUsize) -> Self {
+        let (sender, receiver) = mpsc::unbounded_channel();
+        Self {
+            len: AtomicUsize::new(0),
+            max_timelines_per_batch,
+            sender,
+            receiver: Mutex::new(receiver),
         }
     }
 
-    /// Polls a new task from the queue, using its receiver counterpart.
-    /// Does not block if the queue is empty, returning [`None`] instead.
-    /// Needed to correctly track the queue length.
-    async fn next_task(
-        receiver: &mut UnboundedReceiver<(ZTenantTimelineId, SyncTask)>,
-    ) -> Option<(ZTenantTimelineId, SyncTask)> {
-        let task = receiver.recv().await;
-        if task.is_some() {
-            LENGTH.fetch_sub(1, Ordering::Relaxed);
+    fn push(&self, sync_id: ZTenantTimelineId, new_task: SyncTask) {
+        match self.sender.send((sync_id, new_task)) {
+            Ok(()) => {
+                self.len.fetch_add(1, Ordering::Relaxed);
+            }
+            Err(e) => {
+                error!("failed to push sync task to queue: {e}");
+            }
         }
-        task
     }
 
-    /// Fetches a task batch, not bigger than the given limit.
-    /// Not blocking, can return fewer tasks if the queue does not contain enough.
-    /// Batch tasks are split by timelines, with all related tasks merged into one (download/upload)
-    /// or two (download and upload, if both were found in the queue during batch construction).
-    pub async fn next_task_batch(
-        receiver: &mut UnboundedReceiver<(ZTenantTimelineId, SyncTask)>,
-        max_timelines_to_sync: NonZeroUsize,
-    ) -> ControlFlow<(), HashMap<ZTenantTimelineId, SyncTask>> {
+    /// Fetches a task batch, getting every existing entry from it, grouping by timelines and merging the tasks for every timeline.
+    /// A timeline has to care to not to delete cetain layers from the remote storage before the corresponding uploads happen.
```

Comment:
I can remove that doc line, don't see any ways to enforce with the current model, only to react on various events that happened already.

What we need to decide is how to deal with the local FS deletions during the tasks.
I think it's fine to reread a local metadata and submit attempt to update the index with it, at least now: we have to update the remote json file anyway, why not do that with the updated metadata then, keeping it more consistent.

We need more tests on that case yet.

Reply:
I see the one single problematic case:
If there will be an upload that fails because compaction deleted the file. Will this job update metadata? If the answer is yes what metadata will it be? We need to maintain the invariant of remote consistent lsn which should grow without introducing any gaps. If the metadata we read is newer than the files we actually uploaded - we're in trouble. I think the safest option is to avoid the metadata update entirely. If there is a proper upload task with metadata snapshot taken at the time of a checkpoint then it is ok to update it

Reply:
Yes, that's actually a very good point, I've thought completely oppositely first.
I've moved this comment to the place where we detect such missing file, removed the entire variant branch with this "soft" error.